### PR TITLE
ISLANDORA-1464 Replace pid with object label in solr results.

### PR DESCRIPTION
**Pull request by rosiel**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1464

See also, discussion in : https://jira.duraspace.org/browse/ISLANDORA-1697

# What does this Pull Request do?

Lets you replace a uri or PID in the solr search results display with the label of the object it refers to.

# What's new?

Another admin option in Solr setting > Display Fields (configure). 

# How should this be tested?

Create a display field like RELS_EXT_isMemberOfCollection_uri_ms. notice that it displays the raw URI, e.g. "info:fedora/islandora:root". Then, go into the "Configuration" for that display field and check the box called "Replace PID with Object Label". Click Apply. **Then also click "Save Solr configuration at the bottom of the page.**

Now your collection field will display with the name of the collection.

It should work with PIDs or URIs. It should work with whatever linking you apply to this field, too (though it'd be mighty confusing to link it to the object itself, because the text refers to a different object). And note that if you select "link to a solr search", the behaviour **does not change** when you enable pid-to-object-label rewriting, though the current behaviour is broken due to the slashes in the URI. See https://jira.duraspace.org/browse/ISLANDORA-1749 - link to search result contains error when field contains a slash.

Testing with highlighting could be useful, though I'm not sure what the expected behaviour is, because we're rewriting text that could potentially match with other text.

# Additional Notes:

* Does this change require documentation to be updated?  Yes, if/when this passes a new option will need to be documented.
* Does this change add any new dependencies? No
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)?  No
* Could this change impact execution of existing code? No

# Interested parties
@bondjimbond and @Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by DiegoPino** [comment]
There are some minor coding standard issues. https://travis-ci.org/Islandora/islandora_solr_search/jobs/229569145#L442 . Will be happy to review once they pass. Thanks @rosiel !

**Comment by jonathangreen** [comment]
I'll take a look at this one.

**Comment by rosiel** [comment]
ping! last chance to get this into 1.10 before code freeze on Tuesday Sept 19.

File: includes/utilities.inc
Comment (DiegoPino) [inline includes/utilities.inc]: Just a comment, but this posse (maybe, correct me if wrong) the following issue (and the object load part also). What happens if the current user has no access/is not allowed to see the object that is referred in the PID/URI?  
What happens if the object is actually not there anymore? (orphan who lost parent)

Comment (rosiel) [reply includes/utilities.inc]: Haha want to find out? I am guessing it's not good things. Nice catch, Diego. Thanks!

Comment (rosiel) [reply includes/utilities.inc]: With commit 0790823:  say you have an object appearing in search results (islandora:123) and it contains, in one of its solr fields, a reference to an inaccessible object (secret:999, with label "My secret object"). 

If secret:999 is not in an accessible namespace, and solr is configured to block that namespace, then you should now see "secret:999". If solr allows that namespace because you configured solr inconsistently from islandora, then you see "My secret object".

If secret:999 is blocked by XACML then it should not appear in an islandoraSolrResult (because the filter of 

```RELS_EXT_isViewableByUser_literal_ms:"admin" OR RELS_EXT_isViewableByRole_literal_ms:"authenticated user" OR RELS_EXT_isViewableByRole_literal_ms:"administrator" OR ((*:* -RELS_EXT_isViewableByUser_literal_ms:[* TO *]) AND (*:* -RELS_EXT_isViewableByRole_literal_ms:[* TO *]))``` 

gets added to the islandoraSolrQueryProcessor), so you see "secret:999". 

However, this brings to light another issue: in the Solr Settings > Solr Query Defaults screen, there's a Solr base filter. You might use this to prevent collections or other types of objects from appearing in solr results. The description is "Lists base filters that are appended to all *user* queries" (emphasis mine). That base filter is also applied to these queries, so if you're blocking, say, collections, from appearing in the user's search results, then you're also blocking them from appearing in internal calls such as this. I think this is an issue, so I'll add it to Jira if it's not there already.

Comment (whikloj) [reply includes/utilities.inc]: @rosiel so the issue is that you have (for example)
`resourceA  example:hasRelated resourceB`

I can access `resourceA` but not `resourceB` so by viewing `resourceA` I can see that `resourceB` exists, but you don't want the link to appear at all.

For Solr is sounds like we'd need to know which fields contain a reference to another resource and to loop through those fields after resolving the query to try to filter out the "hidden" links.

Not sure if there is an easy way to do that.

File: includes/utilities.inc
Comment (DiegoPino) [inline includes/utilities.inc]: isset is not the same as !not empty. If empty, you will be returning an empty label

**Comment by DiegoPino** [comment]
@rosiel i see your changes. Seems like most of the issues are addressed, but i don't get this:

> That base filter is also applied to these queries, so if you're blocking, say, collections, from appearing in the user's search results, then you're also blocking them from appearing in internal calls such as this. I think this is an issue, so I'll add it to Jira if it's not there already.

Why is that wrong/and issue? If I actually want to avoid every object that has "diego" in the title to be hidden, why would an internal query like yours be able to override that? I feel it behave the same as the XACML filter and i kinda like the consistency.  Or did i get the issue wrong?

File: includes/utilities.inc
Comment (rosiel) [reply includes/utilities.inc]: @whikloj I was taking a simpler approach: that "example:hasRelated info:fedora/resourceB" is part of resourceA's metadata, and therefore if you can see resourceA you can see that relationship, and resourceB's URI and/or PID. But if you're not allowed to see resourceB then you cannot resolve its URI to its label. 

Does that sound reasonable?

**Comment by rosiel** [comment]
@DiegoPino if that's the case, then this needs to be fixed in documentation. Because "hidden from user queries" and "completely inaccessible to this site" are very different things. Although, this is completely unrelated to this ticket.

Edit: it's already a ticket by @nhart! https://jira.duraspace.org/browse/ISLANDORA-1747
Edit edit: I made a new ticket because I think this is related and also documentation. https://jira.duraspace.org/browse/ISLANDORA-2085

File: includes/utilities.inc
Comment (whikloj) [reply includes/utilities.inc]: Sorry I must have read your comment wrong, I thought you did not like that as the solution. If you are happy, then I'm happy. 😄

**Comment by DiegoPino** [comment]
@rosiel don't feel it is a big concern. Seem like if an "admin/level/tasks" query needs to avoid filters, the query processor itself give you enough flexibility to override fl, qf, etc.[ I remember having to do that manually on some of my modules](https://github.com/DiegoPino/islandora_solution_pack_redbiodiversidad-7.x-dev/blob/94acda21b2573dd1cc37f5999fb031c8cba98730/includes/utilities.inc#L356-L436), and the ticket proposes a query processor "option", but does not says that the current behavior is wrong -wrong- per se. 
Code looks good then, are you ok with it as it is? can we merge? 😄

**Comment by DiegoPino** [comment]
@rosiel ping, i wrote this 15 days ago: "code looks good then, are you ok with it as it is? can we merge? 😄" ? still 😄

**Review by DiegoPino — APPROVED** [review]

**Comment by rosiel** [comment]
I'm hesitant because I don't think it's robust enough. If I can break it just by saying "i don't want that kind of object to clutter up my user's searches" then I think there are too many cases where this fails. If a related object is _structural_ like the name of a collection, a newspaper issue, or a fonds, then it seems likely that's not something you want to be searching on directly. I'm all for incremental improvements driving better software, but I don't want to introduce a feature that's half-broken.

**Comment by bondjimbond** [comment]
@rosiel Based on your last comment, should this PR be closed? Or is it still going to move forward after a few more adjustments?

**Comment by DiegoPino** [comment]
@Islandora/7-x-1-x-committers any decisions on this?

**Comment by rosiel** [comment]
I notice that @whikloj  removes the solr base filters for a similar problem, in a (at this point) open pull request, here: https://github.com/Islandora/islandora_solution_pack_newspaper/pull/158/commits/f7db27b01d048907e2aa230f579646582dfee0ed

If that's an ok solution, I could apply the same thing here.

(Diego, I know you pointed out some code in red_biodiversidad where you did something similar... I'm sorry, it went over my head, and it seemed more like building your own base filters, which is why I didn't implement it)

**Comment by giancarlobi** [comment]
My point of view about this:
- I need this feature because I return also pages as search results and I need to add parent book title other than page number, so I applied @rosiel PR to my Islandora and it works like a charm (thanks @rosiel )
but
- our repository is completely open access so I don't need filter anything
- we use only dismax search
I think this is an important features and very useful for a lot of islandorians.

**Comment by rosiel** [comment]
Mea culpa. I just re-read the code and my previous comment/concern is unnecessary. This PR does just like Facets, and uses Solr with base filters intact, but tries to load the object to get the label if it's not found. I'm ok with it as it is, if y'all are.

**Comment by whikloj** [comment]
@rosiel I think (but I could be wrong) you are rewriting the pid to label for objects that were returned by the first Solr search, so if I as the user didn't have rights to see that object it shouldn't have come up in the results of the search.

**Comment by rosiel** [comment]
@whikloj that is correct... I'm referring to the case where the object whose pid we want to rewrite is excluded from solr queries because of the Solr Base Filter. However, as I just realized, if it doesn't come up from Solr I'll try loading the object directly. If I'm really not allowed to see it (namespace, XACML, or other) I won't be able to get the label even by loading directly. But if it's merely excluded by the solr base filter, I *do* get the label.

**Comment by rosiel** [comment]
ping @DiegoPino @whikloj is any further work needed here?

**Comment by DiegoPino** [comment]
@rosiel not from my part, that is the reason I asked about agreement/ the concerns where yours (self-concerns!) https://github.com/Islandora/islandora_solr_search/pull/318#issuecomment-341548346
So.. let me know, are your own concerns addressed by your commits? 😬 I read that between lines but not 100% sure. If yes, will merge since I approved this just before that concerned message.

**Comment by rosiel** [comment]
Haha yes I think it's OK! I got caught up in the conversation and lost track of how my code actually works.

**Pull request merged**
